### ISSUE TITLE
[enhancement]: prompt2prompt reweight operator

### ISSUE THREAD

**Issue by keturn** [invoke-ai/InvokeAI#4541]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What should this feature add?

Two related things: 1) adding cross-attention controls other than `swap` (including `reweight`), 2) future of maintaining prompt2prompt implementation.

## Background

InvokeAI's implementation of [Prompt-to-Prompt Image Editing with Cross-Attention Control](https://prompt-to-prompt.github.io/) predates Invoke 2.3's adoption of `diffusers`. Often referred to as "prompt2prompt" in implementation discussions, it's documented as [Cross-Attention Control](https://invoke-ai.github.io/InvokeAI/features/PROMPTS/#cross-attention-control-prompt2prompt) in our user docs and the Word Swap control is accessed with the `(foo, bar).swap()` syntax in Invoke prompting.

The implementation has been one of the more complicated things to maintain, as it requires messing with those attention layers inside the U-Net instead of operating on its normal inputs and outputs.

## Recent Developments

diffusers recently accepted a PR for a prompt2prompt pipeline:
- https://github.com/huggingface/diffusers/pull/4563

I was hoping that would mean that Invoke would get to replace most or all of our implementation and use one maintained upstream. I'm also happy to see that it's a more complete implementation of prompt2prompt and includes the Refine control (adding new terms with blending) and Reweight (weighting based on attention, not by scaling the vectors from the text encoder).

Unfortunately diffusers has elected to add it to their `examples` as untested code instead of including it in the distribution, so we don't get out of having to maintain it after all.

# Next Steps

I think it probably makes sense to adopt the code from https://github.com/huggingface/diffusers/pull/4563 into Invoke's codebase, along with the [tests](https://github.com/huggingface/diffusers/blob/6e8afa4fdd255f8d6f606e84afa879ca5cf85fcd/tests/pipelines/prompt2prompt/test_prompt2prompt.py) that were written for it -- but TBH I haven't looked to see how much it differs from our current implementation.

Allowing the use of the other cross-attention control types would require supporting them in compel, or some other user interface.

We should also be thinking about how to keep feature parity between SD and SDXL. Presumably cross-attention controls may be applied to SDXL in a similar manner, but since SDXL's conditioning contains the output of _two_ text encoders, mapping between specific words in the text and the attention layers must be a little more complicated.

**Comment by RyanJDick** [comment]
I've been looking at this as part of the IP-Adapter work (https://github.com/invoke-ai/InvokeAI/pull/4429), since it also overrides the attention blocks. Here are a few thoughts that come to mind:

`diffusers` provides 2 main injection points for features that want to manipulate the cross-attention blocks: [cross_attention_kwargs](https://github.com/huggingface/diffusers/blob/38a664a3d61e27ab18cd698231422b3c38d6eebf/src/diffusers/models/unet_2d_condition.py#L742) and [set_attn_processor](https://github.com/huggingface/diffusers/blob/38a664a3d61e27ab18cd698231422b3c38d6eebf/src/diffusers/models/unet_2d_condition.py#L616). I think it makes sense to stick to these injection points, and avoid monkey-patching the UNet in other ways if possible (as tempting as it is). (This is what we've done with the IP-Adapter implementation and what diffusers appears to have done in the Prompt2Prompt pipeline.)

Related to the above, Cross-Attention Control in InvokeAI currently uses a [custom UNet forward step](https://github.com/invoke-ai/InvokeAI/blob/604fc006b15d9c3f3e5cedbc232e2b4907147f48/invokeai/backend/stable_diffusion/diffusion/shared_invokeai_diffusion.py#L282-L292). I'm pretty sure this could be eliminated entirely and achieved through the injection points mentioned above.

Another thing to keep in mind is that it seems likely that we will have an increasing number of features that want to override the UNet attention blocks. These features will be incompatible with one another (i.e. only one can be enabled at a time). We should plan for this by 1) aiming for a consistent implementation structure for all of them, and 2) thinking about how we will communicate this incompatibility in the UI (I don't have answer for this yet).

**TLDR**
1. Let's use the [cross_attention_kwargs](https://github.com/huggingface/diffusers/blob/38a664a3d61e27ab18cd698231422b3c38d6eebf/src/diffusers/models/unet_2d_condition.py#L742) and [set_attn_processor](https://github.com/huggingface/diffusers/blob/38a664a3d61e27ab18cd698231422b3c38d6eebf/src/diffusers/models/unet_2d_condition.py#L616) injection points, and avoid monkey patching the UNet in other ways when possible.
2. Let's plan to support many features that all override the UNet attention blocks.
